### ISSUE TITLE
how can i include some jade from node_modules?

### ISSUE THREAD

**Issue by i5ting** [jadejs/jade#2164]
example 

```
header.jade in node_modules/xxx/views/header.jade
```

in my express layout 

```
include node_modules/xxx/views/header.jade
```

**Comment by TimothyGu** [comment]
There is no special syntax for files from a npm module. What you have now is correct.

We could consider making a `require` plugin for such things after jade@2 is released though.

**Comment by tjconcept** [comment]
Would be absolutely amazing if you could easily compose stuff using npm.

I think it should follow the same rules as Node.js for resolving. Maybe something like https://www.npmjs.com/package/resolve (developed for Browserify) could be used. Not sure if it will need some patching for `index.js` -> `index.jade`.

**Comment by tomlagier** [comment]
I've written a Gulp plugin that does exactly this! Check it out, maybe it will help: https://github.com/tomlagier/gulp-jade-modules

**Comment by jonscottclark** [comment]
Nice, @tomlagier, looks like that could be helpful.

Having recently been checking out how to accomplish this with Sass, the way `npm-sass` does it is that it defines use a custom "importer" function, which you pass to the "importer" option at compile-time, which extends the default Sass import behaviour to use the closest `node_modules` dir as the first location to check when resolving filenames. Maybe a similar approach could be taken; you could define your own custom "includer" function (in Jade's case) to accomplish this with Jade?

Of course, it wouldn't be obvious to people looking at your code later on that it's not just looking for your includes relative to the current file, which could throw people off. I _definitely_ like the idea of a `require` keyword in jade@2, which I guess would just be an extension of `include`, @TimothyGu :)

**Comment by iliakan** [comment]
is it possible to somehow override jade resolve mechanism?

Like you know, webpack uses `~module/path/to.js` to search in modules. I'd love to do the same. In a modular architecture it's pretty common for one module to use something from another.

**Comment by glebmachine** [comment]
Yeah, it would be awesome to have some props, to define import paths.

Or, maybe, support `node_modules` like:
```jade
include 'bemto/bemto.jade'
```jade
(repo from example: https://github.com/kizu/bemto)

P.S. Worst thing - each time you should write crazy paths like `./../../../node_modules/bemto/bemto.jade`

**Comment by jamespegg** [comment]
I've come across a pretty specific use case where this would be very handy;

In Express 4.1+ you can set multiple view directories. At the moment all that means is that Express first checks the location of the root `.pug` file in one of the defined directories. Once it's found one, it'll just pass that path down to Pug and all the `includes` and `extends` in Pug are in relation to that particular file.

The situation I've got is that I want users to be able to overwrite portions of a base template, which means mixing `includes` and `extends` across a bunch of locations. For example they might want to keep the the nuts and bolts of a particular view, but overwrite the overarching theme template.

There's a couple of ways you could approach this. Firstly, you could build multiple directory resolves directly into Pug (as already suggested). Secondly, you could use the directories defined in Express. It'll need a change in Express to pass that array through (it's a couple of lines to make that happen) and then a bunch of work in Pug to use it. The second option should work in tandem with sort of "options" setting as well.

I'll raise a ticket on the Express repo to see if they'll even consider passing the array of view directories through to view engines.

**Comment by ForbesLindesay** [comment]
I think the basedir option already covers that use case pretty well.

**Comment by jamespegg** [comment]
Ah thanks, I've just been playing around with `basedir` and finally cracked it. In express I've got multiple view directories to load the template but then I've set the "default" directory as pug's `basedir`. 

I couldn't seem to find any mention of `basedir` in the documentation though?

**Comment by TimothyGu** [comment]
The option is documented at https://github.com/pugjs/pug/blob/master/docs/views/api.pug#L36-L38 (since Pug v2 is not released the docs are not up yet).

**Comment by 59naga** [comment]
I was defined the `require` as a custom filter.

```pug
//- index.pug
:require html5-test-page/index.html
```
```js
// pug.config.js
const readFileSync = require('fs').readFileSync

module.exports = {
  filters: {
    require (path) {
      return readFileSync(require.resolve(path), 'utf8')
    }
  }
}
```
```bash
npm install pug-cli html5-test-page
pug index.pug --obj pug.config.js
# rendered index.html
```

**Comment by admdh** [comment]
it would be great to implement something like postcss-import plugin:
```
This plugin can consume local files, node modules or web_modules. To resolve path of an @import rule, it can look into root directory (by default process.cwd()), web_modules, node_modules or local modules. When importing a module, it will looks for index.css or file referenced in package.json in the style or main fields. You can also provide manually multiples paths where to look at.
```
https://github.com/postcss/postcss-import


so in your node plugins package.json you could add for example something like this:
```json
"pug-file": "index.pug"
```
in your projects package.json just include that plugin as a dependency,
and than just import ```index.pug``` wherever you need for example in some other index.pug

**Comment by puzrin** [comment]
> I think the basedir option already covers that use case pretty well.

Only partially IMHO. npm v2 and npmv3 organize dependencies differently. That does not make sence for root application. But dependencies will have different location of their submodules.

For example, i have doc generator, that needs `include jquery/jquery.js` in template. If i install it standalone, then `app_root/node_modules` is known. But if i install it as dependency, i should traverse up to find folder with jquery.

**Comment by ForbesLindesay** [comment]
I was referring specifically to @jamespegg's use case.  There is still definitely a use case for something that lets you include from/extend from npm modules.

**Comment by puzrin** [comment]
@ForbesLindesay probably i did not found a proper ticket, sorry. Is there one related to supporting npm resolve strategy in `include`? I currently monkey-patch jade for such behaviour, but it would be nice to have something native (or plugin) in pug 2.+.

**Comment by ForbesLindesay** [comment]
There will probably be a plugin at some point for pug 2 (you could even write your own).  This is the correct issue, but my earlier comment was referring directly to the comment before it, not the issue as a whole.

**Comment by iliakan** [comment]
So now it's difficult to make such plugin? I was inheriting from jade and patching this behavior.

**Comment by ForbesLindesay** [comment]
It shouldn't be difficult with the latest pug version, just build a plugin that looks something like:

```js
const path = require('path');
const resolve = require('resolve');

function npmResolverPlugin() {
  return {
    resolve(filename, source, options) {
      return resolve.sync(filename, {basedir: path.dirname(source)});
    }
  };
}
```

Then pass the following options to pug:

```js
{plugins: [npmResolverPlugin()]}
```

Note that only one plugin at a time can override the `resolve` function.

**Comment by iliakan** [comment]
Wow. Looks marvellous! @puzrin?

**Comment by puzrin** [comment]
The only question i have - is 2.0 beta stable enougth for production :)

**Comment by ForbesLindesay** [comment]
@puzrin Yes

**Comment by markbrouch** [comment]
For anyone else stumbling across this thread - I went ahead and made a plugin that does this which you can install from npm: https://github.com/markbrouch/pug-plugin-modules

Please help me test it!
